From a857508425cac80fd1c49a85f4c525d6ff77ce52 Mon Sep 17 00:00:00 2001
From: Rufus Hamade <rufus.hamade@imgtec.com>
Date: Tue, 5 Jul 2016 16:35:19 +0100
Subject: [PATCH] pvr/dri: Pass buffer source to CreateFromFd

The IMG DDK's implementation of DMA_buf is not complete.  In particular,
DMA_buf objects allocated via the DDK do not have an associated
reservation object.  This is mostly not a problem as we normally don't
export DMA_buf objects allocated by the DDK.  Synchronisation on these
objects was done using internal DDK mechanisms.

Unfortunately, the Android PBuffer implementation uses DMA_buf objects
allocated by the DDK.  To deal with this, we  we need to tell the DDK
whether the DMA_buf object is coming in via the image extension (and is
therefore externally allocated) or via the dri extension (and therefore
internally-allocated.)

In future we will implement full support for the DMA_buf API in
DDK-allocated buffers.  When this happens, this change can be removed.

Change-Id: Ie0c98a101cfd514d6a10aa2b8be40c97542aecfc
Signed-off-by: Rufus Hamade <rufus.hamade@imgtec.com>
---
 src/mesa/drivers/dri/pvr/pvr_dri_support.h | 3 ++-
 src/mesa/drivers/dri/pvr/pvrimage.c        | 3 ++-
 2 files changed, 4 insertions(+), 2 deletions(-)

diff --git a/src/mesa/drivers/dri/pvr/pvr_dri_support.h b/src/mesa/drivers/dri/pvr/pvr_dri_support.h
index fb26b12..f996a3d 100644
--- a/src/mesa/drivers/dri/pvr/pvr_dri_support.h
+++ b/src/mesa/drivers/dri/pvr/pvr_dri_support.h
@@ -250,7 +250,8 @@ PVRDRIBufferImpl *PVRDRIBufferCreateFromFd(PVRDRIScreenImpl *psScreenImpl,
 					   int iWidth,
 					   int iHeight,
 					   unsigned int uiStride,
-					   unsigned int uiOffset);
+					   unsigned int uiOffset,
+					   bool isInternallyAllocated);
 
 void PVRDRIBufferDestroy(PVRDRIBufferImpl *psBuffer);
 
diff --git a/src/mesa/drivers/dri/pvr/pvrimage.c b/src/mesa/drivers/dri/pvr/pvrimage.c
index f912cc4..368ebc4 100644
--- a/src/mesa/drivers/dri/pvr/pvrimage.c
+++ b/src/mesa/drivers/dri/pvr/pvrimage.c
@@ -295,7 +295,8 @@ CreateImageSharedFromDmaBufs(__DRIscreen *screen,
 						 width >> psFormat->sPlanes[i].uiWidthShift,
 						 height >> psFormat->sPlanes[i].uiHeightShift,
 						 strides[i],
-						 offsets[i]);
+						 offsets[i],
+						 false);
 		if (!shared->apsBuffers[i])
 		{
 			errorMessage("%s: Failed to create plane %d for shared image\n", __func__, i);
-- 
1.9.1

